fms/modules/feedforward.py

```diff
@@ -326,65 +369,33 @@ def __init__(
-        used_keys: Set[str] = set()
-        if self.fused:
-            wg_weight = self._get_sd_weight(
-                tensor_values, used_keys, ["wg1_fused", "weight"]
-            )
-        else:
-            w1_weight = self._get_sd_weight(tensor_values, used_keys, ["w1", "weight"])
-            wg_weight = self._get_sd_weight(tensor_values, used_keys, ["wg", "weight"])
-        w2_weight = self._get_sd_weight(tensor_values, used_keys, ["w2", "weight"])
-        if self.use_bias:
-            if self.fused:
-                wg_bias = self._get_sd_weight(
-                    tensor_values, used_keys, ["wg1_fused", "bias"]
-                )
-            else:
-                w1_bias = self._get_sd_weight(tensor_values, used_keys, ["w1", "bias"])
-                wg_bias = self._get_sd_weight(tensor_values, used_keys, ["wg", "bias"])
-            w2_bias = self._get_sd_weight(tensor_values, used_keys, ["w2", "bias"])
 
-        # 2. Raise exceptions
-        num_weights = 2 if self.fused else 3
-        if len(tensor_values) > (num_weights * 2 if self.use_bias else num_weights):
-            unused_keys = set(tensor_values.keys()).difference(used_keys)
-            raise AttributeError(f"Unused weight(s): {', '.join(unused_keys)}")
-
-        # 3. Load and shard the weights
+        # sharding modules struct: {'module_name': (module_obj, sharding_dim, max_partition)}
         if self.fused:
-            self.sharded_copy(
-                self.wg1_fused.weight, wg_weight, 0, [self.world_size, self.world_size]
-            )
+            module_sharding_info = {
+                "wg1_fused": LinearModuleShardingInfo(self.wg1_fused, 0, [self.world_size, self.world_size]),
+            }
         else:
-            self.sharded_copy(
-                self.w1.weight, w1_weight, 0, [self.world_size, self.world_size]
-            )
-            self.sharded_copy(
-                self.wg.weight, wg_weight, 0, [self.world_size, self.world_size]
-            )
-        self.sharded_copy(self.w2.weight, w2_weight, 1, [self.world_size])
-        if self.use_bias:
-            if self.fused:
-                self.sharded_copy(
-                    self.wg1_fused.bias, wg_bias, 0, [self.world_size, self.world_size]
-                )
-            else:
-                self.sharded_copy(
-                    self.w1.bias, w1_bias, 0, [self.world_size, self.world_size]
-                )
-                self.sharded_copy(
-                    self.wg.bias, wg_bias, 0, [self.world_size, self.world_size]
-                )
-            self.sharded_copy(self.w2.bias, w2_bias, 1, [self.world_size], False)
+            module_sharding_info = {
+                "w1": LinearModuleShardingInfo(self.w1, 0, [self.world_size]),
+                "wg": LinearModuleShardingInfo(self.wg, 0, [self.world_size]),
+            }
+        module_sharding_info.update({"w2": LinearModuleShardingInfo(self.w2, 1, [self.world_size])})
```

Comment:
why use `update` instead of the more pythonic `module_sharding_info["w2"] = LinearSahrdingInfo(...)`?